### PR TITLE
feat(module:watermark): add watermark component

### PR DESCRIPTION
**Pull request by OriginRing**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: components/water-mark/demo/custom.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  content: string = 'NG Ant Design';
+  color: string = 'rgba(0,0,0,.15)';
+  font: FontType = {
+    color: 'rgba(0,0,0,.15)',
+    fontSize: 16
+  };
+  zIndex: number = 11;
+  rotate: number = -22;
+  gap: [number, number] = [100, 100];
+  offset!: [number, number];
```
Comment (hsuanxyz) [inline components/water-mark/demo/custom.ts:152]:
```suggestion
  offset: [number, number];
```

File: components/water-mark/water-mark.component.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+ */
+const BaseSize = 2;
+const FontGap = 3;
+
+@Component({
+  selector: 'nz-water-mark',
+  exportAs: 'NzWaterMark',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  template: `
+    <div style="position: relative;" #watermark>
```
Comment (hsuanxyz) [inline components/water-mark/water-mark.component.ts:35]: 直接写在 host 上

File: components/water-mark/water-mark.component.ts [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+  @Input() nzHeight: number = 64;
+  @Input() nzRotate: number = -22;
+  @Input() nzZIndex: number = 9;
+  @Input() nzImage: string = '';
+  @Input() nzContent: string | string[] = '';
+  @Input() nzFont: FontType = {};
+  @Input() nzGap: [number, number] = [100, 100];
+  @Input() nzOffset: [number, number] = [this.nzGap[0] / 2, this.nzGap[1] / 2];
+
+  @ViewChild('watermark', { static: false }) containerElement!: ElementRef<HTMLDivElement>;
```
Comment (hsuanxyz) [inline components/water-mark/water-mark.component.ts:51]: 直接用 host

File: components/water-mark/water-mark.component.ts [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+  waterMarkElement: HTMLDivElement = document.createElement('div');
```
Comment (hsuanxyz) [inline components/water-mark/water-mark.component.ts:52]: 用  `import { DOCUMENT } from '@angular/common';`

**Review by hsuanxyz — CHANGES_REQUESTED** [review]

**Review by simplejason — APPROVED** [review]
LGTM

**Pull request merged**